### DISCUSSION TITLE
Add Client implementation

### DISCUSSION THREAD

**Discussion by Nailik** [davidepianca98/KMQTT#16] [Ideas]
Hi,

i am searching for a MQTT Client library for KMM.

Mostly because:

- just on implementation for iOS and Android
- all the Android Client libraries lag the possibility to use a inputStream to send bigger data (because when loading all the data in memory first there is a possibility of an OutOfMemory exception which currently is an issue in one of my projects)

Do you think adding Client implementation to this library or creating a new library is better?

**Comment by davidepianca98** [comment]
Hello, implementing the client side is definitely something I would like to do, especially if there is enough interest in the project. A few things are already implemented, like the packets, so it's mainly the logic missing. I might consider working on it but it's not something that would be ready in a few days unfortunately. Which MQTT version would you need first?

**Reply by Nailik** [reply]
I think most of my users currently use MQTT 3 (or 3.1.1) .
And for my need (using streams to send/receive data) i think there would also be the need to change some things in the current implementation because currently everything first reads the data into memory.

**Reply by Nailik** [reply]
and also iOS is currently not supported right?

**Reply by davidepianca98** [reply]
It would definitely be easier to make it stream based in the client, while it is not necessary on the server unless you specifically need it there also. Modifying the server logic would be harder though, but it would be nice to do it one day.
iOS is not supported because it doesn't make much sense to host a MQTT broker on a mobile phone. If I were to implement the client I'd like to make it work on many targets like jvm, js, linuxX64, iosArm64 and all the others in tier 1 and tier 2 here https://kotlinlang.org/docs/native-target-support.html.

**Reply by Nailik** [reply]
when i got some free time on the weekend i will start a client implementation. if you want we could talk about architecture first together.

**Reply by davidepianca98** [reply]
I will have a look at the needed project structure modifications to build client and server separately this afternoon. Then I'll see if I can make a quick basic JVM implementation so we have a place where to start.

**Reply by davidepianca98** [reply]
I have completed a basic generic client implementation. You can find it on the `client` branch. I have only tried it on the JVM, but tomorrow I will try it on posix targets, then once it works there it should be easy to make it work on iOS (after building OpenSSL) and then we can try to refine it and make the API look better.

**Reply by davidepianca98** [reply]
The client seems to be working on posix targets. I was trying to understand how to build OpenSSL for iOS but it looks like it's only possible from an Apple computer, which I don't have. This is how I build it for the other targets if you want to try: https://github.com/davidepianca98/KMQTT/blob/client/kmqtt-broker/src/linuxX64Main/resources/Dockerfile. I have already implemented the platform specific functions for iOS, so only OpenSSL is missing.

**Reply by Nailik** [reply]
Do i see correctly that client is already working since Mar 22?

**Reply by Nailik** [reply]
However could you publish it to jitpack?

**Reply by davidepianca98** [reply]
Hello, the client is available for pretty much any Kotlin Multiplatform target now. JitPack uses Linux to build the artifacts, but Apple doesn't allow cross compilation, so you specifically need a MacOS device to build those artifacts. Consequentially JitPack cannot build those artifacts. This has been a known issue for many years and it probably will never be solved. So the artifacts are now published on GitHub packages. The updated instructions are in the Readme, please take a look https://github.com/davidepianca98/KMQTT/blob/master/Readme.md.

**Reply by Nailik** [reply]
Nice thank you. In the following weeks i'll make a pull request that uses a kotlinx-io source object to send data via mqtt because this would enable to send bigger files without reading them into memory at once.
However kotlinx-io still needs some features that are missing, that's why i am waiting at the moment.

**Reply by Nailik** [reply]
Also MQTT Version should be set as enum because it is really confusing that 4 = 3.1.1 and 5 = 5

**Reply by davidepianca98** [reply]
> Also MQTT Version should be set as enum because it is really confusing that 4 = 3.1.1 and 5 = 5

Thank you for the suggestion, I have done that now. I know the numbers are confusing but that's the specification, there's not much we can do about that.
